Add async version of the accountability algorithm, support for multi-height experiments, monitor initialization through config file and optional report generation

mikanikos opened the pull request:
Closes #5, closes #11, closes #7, closes #10, closes #9, closes #14

jovankomatovic5 commented on cmd/monitor/monitor.go:
```diff
@@ -0,0 +1,179 @@
+		go monitor.receiveHvsFromValidator(conn, validatorCloseChannel)
+
+		// periodic send request to validator
+		go conn.PeriodicSend(packet, validatorCloseChannel, sendTimer)
```
Monitor does not need to periodically request Height-Vote-Set, since we assume the reliable links.

mikanikos replied:
Yes I know that the model assumes perfect links between processes but real systems can have failures of any kind (process failures, network failures etc.) and the idea was to handle more case scenarios. Nevertheless, if the processes are correct and the network is reliable, the implementation behaves overall as the monitor algorithm specs describe; on the other hand, if there are some failures (some validator not responding back to the request, monitor not receiving some packets etc.) the monitor tries to handle the situation in the best way possible, always following the algorithm. We can better discuss this in the next meeting though and I can show you my reasoning.

jovankomatovic5 replied:
I agree with you that it makes more sense to use this logic in a real-systems scenario.

jovankomatovic5 commented on accountability/accountability.go:
```diff
@@ -1,45 +1,58 @@
+			// if process didn't send the hvs, just ignore
 			if hvs == nil || !hvsLoad {
-				faultyProcesses.AddFaultinessReason(NewFaultiness(processIndex, 0, ErrHVSNotSent))
+				//acc.FaultySet.AddFaultinessReason(NewFaultiness(processIndex, 0, FaultinessHVSNotSent))
```
Even if monitor did not receive the Height-Vote-Set of some process P, it must execute this.
Why?
Because, monitor may notice equivocation and will be able to expose some processes.

Let me give you an example.
Suppose that you have that two correct process P and Q decide differently in round R.
Hence, there is an equivocation.
Moreover, suppose that monitor receives Height-Vote-Sets from f + 1 correct guys (hence, it does not receive Height-Vote-Sets of faulty processes that committed equivocation).
By ignoring them (since monitor did not receive their Height-Vote-Sets), monitor will not be able to expose them (and note that these faulty guys will never send their Height-Vote-Sets to monitor).

Please, let me know if you cover this scenario, but I did not understand it.

mikanikos replied:
That's perfectly true, I removed that line purposely because I thought that the new async version couldn't handle some cases. But I'm still not sure if adding that line would solve all problems. For instance, what happens in the opposite scenario?

Imagine the same example you described but the monitor receives Height-Vote-Sets from f + 1 faulty guys first (correct validators are slow). The monitor runs the algorithm because it received f+1 message logs: since A and B didn't send their logs, the monitor will detect them and the algorithm completes by wrongly stating that A and B are faulty.

It's likely that I'm wrong and I'm not considering something important, so let me know what I'm missing. By the way, I'm mostly relying on the document you sent me by email, I don't know if you updated it or something changed.

jovankomatovic5 replied:
I am not sure I understand which processes A and B are?
Note the following, we assume that messages can not be forged, i.e., some process P can claim that it has received message M from process Q only if Q indeed sent this message.

mikanikos replied:
Sorry I modified my previous comment: I was writing a different example, then changed my mind but forgot to update the last part. What I'm saying is that the monitor can wrongly detect some processes only because they have been slow to send their logs: since the algorithm is run once the monitor has received f+1 message logs, if we detect all the other processes that didn't send their logs yet the algorithm would not work correctly.

jovankomatovic5 replied:
Not really, that's the beauty of it.
Monitor will run it for the first time once it delivers f + 1 Height-Vote-Sets and will try to expose f + 1 processes. If monitor does not succeed, it will wait for new Height-Vote-Sets.

The key here is to understand the following: Once monitor receives Height-Vote-Sets from all correct processes, it will be able to expose at least f + 1 faulty processes.
Hence, if monitor does not succeed in exposing f + 1 processes in the first "run", it can be sure that it will receive more Height-Vote-Sets.
Did I clarify things now?

mikanikos replied:
Yes I know that the monitor will wait for new HVSs if it fails detecting f+1 processes but, in this case, it will expose f+1 processes just because they didn't send their HVSs. I'll give you the example I was writing previously which maybe it's more clear:

Suppose we have 4 processes, two are correct (A and B) and two are faulty (C and D).
The monitor receives the HVSs from C and D (assume A and B are slow): since the monitor received f+1 HVSs, it runs the algorithm and it will detect A and B by executing that line above, just because it didn't receive the HVSs of A and B yet. Since the monitor exposed f+1 processes (A and B), the monitor stops. Please let me know if it's more clear now.

jovankomatovic5 replied:
Yes, but you did not understand my comment.

So, monitor will not expose them because they did not send their Height-Vote-Set. It will just add messages C and D received from A or B to Height-Vote-Sets of A and B (that are empty since monitor did not receive them).
Then, monitor will run through Height-Vote-Sets of A and try to find equivocation and it will not succeed since A is correct (same for B).
Bottom line is, monitor does not expose some process P if it did not receive its Height-Vote-Set, but only if it figures that P equivocated.
You see the difference?

mikanikos replied:
You're right, I see the difference now! However, adding the above line would not be completely correct, it must be added when we check if a process is faulty and not during the pre-processing phase. Thanks a lot for that, I'll fix this.

mikanikos replied:
To better explain the problem to whoever is gonna read this, adding the above line would be wrong because we are just ignoring processes from the pre-processing phase, not from the faultiness-detection algorithm. There's just one small bug few lines later that must be fixed because the monitor should initialize empty Height-Vote-Sets even if it didn't receive anything from some processes.

mikanikos replied:
Actually, the last fix we discussed yesterday during the call is not enough and generated another problem. We can discuss it later in the scheduled call.

jovankomatovic5 replied:
Can you briefly explain the problem here?
Thanks :)

mikanikos replied:
As we know, the implemented algorithm has two consecutive phases: pre-processing phase and fault-detection phase.

In the pre-processing phase, we go through all the received height vote sets so far and we "infer" the missing height vote sets of the processes that didn't send their logs yet. However, during this phase, we only infer the sent messages by the processes that didn't send their logs based on the received height vote sets that the monitor collected.

When it comes to the second phase, we run the algorithm to check whether a process is faulty or not (based on its height vote set, either received or inferred) but we cannot use the inferred height vote sets to check for a faultiness different than equivocation. 

In fact, when checking if a process received enough (2f+1) prevotes to send a precommit (or enough prevotes to send another prevote in a later round), the process with the inferred height vote set will not contain any message in its received messages bucket because the pre-processing phase infers only the messages that have been sent by the process. Therefore, the process will always result faulty because it doesn't have enough prevotes to justify a sent precommit (or prevote) since it doesn't have any received messages in any round.

As far as I understood, the inferred height vote sets should only be checked for equivocation (i.e. a process sent two or more messages with the same type in the same round but with different value). Would this assumption be correct?

Please let me know if everything is clear.

jovankomatovic5 replied:
I understood, thanks! :)
However, even in these cases the monitor will make a correct call (that is why for asynchronous cases some messages need to contain justification).
So, the fix we had yesterday is ok.

I will explain in more details everything today.
Please, do not change the code before our meeting, so we can discuss this scenario,
Thanks!

mikanikos replied:
I've already changed and refactored the code yesterday but we can start from this commit https://github.com/mikanikos/Fork-Accountability/tree/e3327e0fd55e40156fff30dfdbac068d6fa05585.

mikanikos replied:
As we said, inferred height vote sets should only be checked for equivocation. Fixed in the last commit.

mikanikos commented on accountability/accountability.go:
```diff
@@ -0,0 +1,213 @@
+	"sync"
+)
+
+// Accountability stores all the validators that are faulty and the corresponding faultiness proofs
```
I'm not sure about this. For example, when we say "Faultiness Reason" we mean a reason for the faultiness found (i.e. a reason for being categorized as faulty) but "Faulty Reason" would mean that the reason is faulty, which is not really the same imho. Of course, my very poor level of English is not the most reliable source to argue this. In any case, that's a minor thing but if you prefer I can change it.

mikanikos commented on accountability/heightlogs.go:
```diff
@@ -0,0 +1,63 @@
+func (hl *HeightLogs) AddHvs(processID string, hvs *common.HeightVoteSet) {
+	hl.mutex.Lock()
+	defer hl.mutex.Unlock()
+	hl.logs[processID] = hvs
```
Yes, it is possible that the hvs already exists but this can only happen if the validators configs are initialized with non-unique ids. However, I was still handling this case by checking whether the map already contained the same value (there was a Contains method for that) and, if yes, I was not considering the hvs received (by the way, that's one of the many cases where a timeout for the monitor could be useful: if a wrong or non-valid hvs is sent, the algorithm might not be able to complete despite the theoretical correctness). Now, I merged the two methods for simplicity and use a LoadOrStore but the behavior is the same and should not give problems.

mikanikos commented on cmd/monitor/main.go:
```diff
@@ -0,0 +1,35 @@
+	flag.Parse()
+
+	// parse file
+	monitor, err := parseMonitorConfig(*configFile)
```
Good point! Refactored in the last commit.

mikanikos commented on connection/config.go:
```diff
@@ -4,8 +4,9 @@ import "github.com/mikanikos/Fork-Accountability/common"
 type Packet struct {
-	Code uint32
-	Hvs  *common.HeightVoteSet
+	Code   uint32
```
A code field in the Packet is simply a way to categorize the packets and regulate the communication between monitor and validators. The monitor always sends a packet with code "HvsRequest" to indicate it's requesting a hvs from validators. Validators can answer with code HvsResponse or with code HvsMissing to notify the monitor they don't have the hvs requested for a specific height.

mikanikos commented:
I refactored the project and added the report feature (closes #14) since I've already started working on it in the last few days. I've also been particularly careful to modify the accountability algorithm in order to send justifications in all the cases we discussed during the last call. For double-checking, you can have a look at it https://github.com/mikanikos/Fork-Accountability/blob/feature/async-version/accountability/algorithm.go, in particular from line 95 to 120 and the method at line 154 which contain the small modifications I made.

Regarding the monitor timeout thing, I would still prefer to keep it to prevent a "wait-forever" state in case of some implementation problems (network issue, invalid hvs, error in config files etc.). As I said, it can be easily modified from the monitor config file and currently is set to 60 seconds, which should be enough for every medium/large experiment. So, we can just set it to a very high value and pretend it doesn't even exist :) Of course, if you insist to remove it I'll do it without problems.

Apart from these, if you have any other suggestion or concern please feel free to let me know. I'll merge the PR at the end of the week if there are no other remarks from your side and I'll start working on the formalization and theoretical description of the algorithm and the context of the project as we agreed in the last call.
